File: packages/innerr-attribution/resources/android/AttributionFunctions.kt
package app.innerr.attribution

import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.provider.Settings
import android.util.Log
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.singular.sdk.Singular
import com.singular.sdk.SingularConfig
import org.json.JSONObject

/**
 * Privacy-first mobile attribution via the Singular SDK.
 *
 * Android attributes through the Google Play Install Referrer; no advertising id
 * (GAID) is collected because the plugin deliberately does NOT declare the
 * com.google.android.gms.permission.AD_ID permission. This keeps measurement in
 * line with the app's privacy positioning while still attributing paid installs.
 *
 * Singular's SDK key + secret are client-side keys (embedded in every install by
 * design). They are injected at build time from .env via the plugin manifest's
 * `secrets` + `meta_data` placeholders and read here from the app's <meta-data>.
 *
 * NOTE: verify the SDK method signatures against the pinned version at
 * developers.singular.net — Singular's API evolves between major versions.
 *
 * Namespace: "Attribution.*"
 */
object AttributionFunctions {

    private const val TAG = "Attribution"
    private const val META_API_KEY = "app.innerr.attribution.SINGULAR_API_KEY"
    private const val META_SECRET = "app.innerr.attribution.SINGULAR_SECRET"

    @Volatile
    private var initialised = false

    /**
     * Log to Logcat only on debuggable builds, so release builds stay silent.
     * Never logs key/secret values.
     */
    private fun debugLog(activity: FragmentActivity, message: String) {
        val debuggable = (activity.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
        if (debuggable) {
            Log.d(TAG, message)
        }
    }

    class Init(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            if (initialised) {
                debugLog(activity, "init: already initialised")
                return BridgeResponse.success(mapOf("success" to true))
            }

            val metaData = activity.packageManager
                .getApplicationInfo(activity.packageName, PackageManager.GET_META_DATA)
                .metaData
            val apiKey = metaData?.getString(META_API_KEY)
            val secret = metaData?.getString(META_SECRET)

            if (apiKey.isNullOrBlank() || secret.isNullOrBlank()) {
                debugLog(activity, "init: SINGULAR_API_KEY/SINGULAR_SECRET missing in <meta-data>")
                return BridgeResponse.error(
                    "SINGULAR_KEYS_MISSING",
                    "Singular keys missing; set SINGULAR_API_KEY and SINGULAR_SECRET in .env",
                )
            }

            val config = SingularConfig(apiKey, secret)
                // Limit downstream sharing of attribution data (privacy posture).
                .withLimitDataSharing(true)

            Singular.init(activity.applicationContext, config)
            initialised = true
            debugLog(activity, "init: Singular initialised (SKAN-only, no IDFA/ATT)")

            // ANDROID_ID is a non-advertising device id; safe to log in debug.
            // Register it as a test device in Singular's Testing Console.
            val androidId = Settings.Secure.getString(
                activity.contentResolver,
                Settings.Secure.ANDROID_ID,
            )
            debugLog(activity, "test device ANDROID_ID: $androidId")

            return BridgeResponse.success(mapOf("success" to true))
        }
    }

    class Event(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val name = parameters["name"] as? String
            if (name == null) {
                debugLog(activity, "event: missing name")
                return BridgeResponse.error("INVALID_PARAMETERS", "Attribution.Event requires a 'name'")
            }

            val value = (parameters["value"] as? Number)?.toDouble()
            val currency = parameters["currency"] as? String

            @Suppress("UNCHECKED_CAST")
            val attributes = parameters["attributes"] as? Map<String, Any>

            when {
                value != null && currency != null -> Singular.customRevenue(name, currency, value)
                attributes != null -> Singular.eventJSON(name, JSONObject(attributes))
                else -> Singular.event(name)
            }

            debugLog(activity, "event: $name")

            return BridgeResponse.success(mapOf("success" to true))
        }
    }

    class SetUserId(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val userId = parameters["userId"] as? String
            if (userId == null) {
                debugLog(activity, "setUserId: missing userId")
                return BridgeResponse.error("INVALID_PARAMETERS", "Attribution.SetUserId requires a 'userId'")
            }

            Singular.setCustomUserId(userId)
            debugLog(activity, "setUserId: set (${userId.length} chars)")

            return BridgeResponse.success(mapOf("success" to true))
        }
    }
}

File: packages/innerr-attribution/resources/ios/AttributionFunctions.swift
import Foundation
import Singular
import UIKit

#if canImport(AdServices)
    import AdServices
#endif

// MARK: - Attribution Function Namespace

/// Privacy-first mobile attribution via the Singular SDK.
///
/// Runs SKAdNetwork-only: `skAdNetworkEnabled` is on and Singular manages the
/// conversion-value model, while no IDFA is requested and no App Tracking
/// Transparency prompt is shown. This matches the app's privacy positioning.
/// Apple Search Ads is attributed separately and deterministically through
/// Apple's own AdServices framework, so nothing is needed here for that channel.
///
/// Singular's SDK key + secret are client-side keys (embedded in every install
/// by design). They are injected at build time from .env into Info.plist
/// (SingularApiKey / SingularSecret) via the plugin manifest and read at init.
///
/// NOTE: verify the SDK method signatures against the pinned pod version at
/// developers.singular.net — Singular's API evolves between major versions.
///
/// Namespace: "Attribution.*"
enum AttributionFunctions {

    private static var initialised = false

    /// Log to the console only on DEBUG builds, so release builds stay silent.
    /// Never logs key/secret values.
    private static func debugLog(_ message: String) {
        #if DEBUG
            // NSLog surfaces in Console.app and the Xcode console; print() only
            // reaches stdout, which `native:run` often does not capture.
            NSLog("[Attribution] %@", message)
        #endif
    }

    /// Initialise the Singular SDK (SKAdNetwork-only, no IDFA/ATT).
    /// Returns: success: boolean
    class Init: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard !AttributionFunctions.initialised else {
                AttributionFunctions.debugLog("init: already initialised")
                return ["success": true]
            }

            guard let apiKey = Bundle.main.object(forInfoDictionaryKey: "SingularApiKey") as? String,
                  let secret = Bundle.main.object(forInfoDictionaryKey: "SingularSecret") as? String,
                  !apiKey.isEmpty, !secret.isEmpty,
                  let config = SingularConfig(apiKey: apiKey, andSecret: secret) else {
                AttributionFunctions.debugLog("init: SingularApiKey/SingularSecret missing or invalid in Info.plist")
                return ["success": false]
            }

            // SKAdNetwork on, Singular-managed conversion values, no ATT wait.
            config.skAdNetworkEnabled = true
            config.manualSkanConversionManagement = false

            #if DEBUG
                // Verbose SDK logging: prints the Singular Device ID and the
                // session/event API requests + responses to the console.
                config.enableLogging = true
            #endif

            Singular.start(config)
            AttributionFunctions.initialised = true
            AttributionFunctions.debugLog("init: Singular started (SKAN-only, no IDFA/ATT)")

            // IDFV is a vendor-scoped, non-advertising id; safe to log in debug.
            // Register it as a test device in Singular's Testing Console.
            if let idfv = UIDevice.current.identifierForVendor?.uuidString {
                AttributionFunctions.debugLog("test device IDFV: \(idfv)")
            }

            #if canImport(AdServices)
                // Reference AdServices so the linker keeps it in EVERY build config
                // (not just DEBUG) — Singular reads the Apple Search Ads attribution
                // token (AAAttribution) from it for deterministic, no-IDFA ASA
                // attribution. There is no manifest key for system frameworks, so we
                // force the link here. Min iOS 18.2 always has AdServices.
                _ = AAAttribution.self

                #if DEBUG
                    if let token = try? AAAttribution.attributionToken() {
                        AttributionFunctions.debugLog(
                            "AdServices token OK (\(token.count) chars) — Apple Search Ads attribution available")
                    } else {
                        AttributionFunctions.debugLog(
                            "AdServices token unavailable (normal on simulator)")
                    }
                #endif
            #endif

            return ["success": true]
        }
    }

    /// Track a conversion event.
    /// Parameters:
    ///   - name: string (required) - prefer Singular standard event tokens
    ///   - value: (optional) double - monetary value
    ///   - currency: (optional) string - ISO 4217, required with value
    ///   - attributes: (optional) [String: Any] - non-PII extras
    /// Returns: success: boolean
    class Event: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let name = parameters["name"] as? String else {
                AttributionFunctions.debugLog("event: missing name")
                return ["success": false]
            }

            let value = parameters["value"] as? Double
            let currency = parameters["currency"] as? String
            let attributes = parameters["attributes"] as? [String: Any]

            if let value, let currency {
                Singular.customRevenue(name, currency: currency, amount: value)
            } else if let attributes {
                Singular.event(name, withArgs: attributes)
            } else {
                Singular.event(name)
            }

            AttributionFunctions.debugLog("event: \(name)")

            return ["success": true]
        }
    }

    /// Associate a hashed first-party user id (never PII).
    /// Parameters:
    ///   - userId: string (required)
    /// Returns: success: boolean
    class SetUserId: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let userId = parameters["userId"] as? String else {
                AttributionFunctions.debugLog("setUserId: missing userId")
                return ["success": false]
            }

            Singular.setCustomUserId(userId)
            AttributionFunctions.debugLog("setUserId: set (\(userId.count) chars)")

            return ["success": true]
        }
    }
}
